src/lib/manager_impl.cpp

```diff
@@ -354,23 +354,57 @@ bool ManagerImpl::bind_data_service_request(std::string const& request_name, gro
     return _mesg_service->bind_data_service_request(request_name, group_id, request_handler);
 }
 
-NullAsyncResult repl_service_ctx_grpc::data_service_request_unidirectional(destination_t, std::string const&,
-                                                                           io_blob_list_t const&) {
-    LOGW("method not implemented");
-    return folly::makeUnexpected(nuraft::cmd_result_code::CANCELLED);
+// The endpoint field of the raft_server config is the uuid of the server.
+const std::string repl_service_ctx_grpc::id_to_str(int32_t const id) const {
+    auto const& srv_config = m_server->raft_server()->get_config()->get_server(id);
+    return (srv_config) ? srv_config->get_endpoint() : std::string();
 }
 
-AsyncResult< sisl::io_blob > repl_service_ctx_grpc::data_service_request_bidrectional(destination_t, std::string const&,
-                                                                                      io_blob_list_t const&) {
-    LOGW("method not implemented");
-    return folly::makeUnexpected(nuraft::cmd_result_code::CANCELLED);
+const std::optional< peer_id_t > repl_service_ctx_grpc::get_peer_id(destination_t const& dest) const {
+    if (std::holds_alternative< peer_id_t >(dest)) return std::get< peer_id_t >(dest);
+    if (std::holds_alternative< role_regex >(dest)) {
+        if (!m_server) {
+            LOGW("server not initialized");
+            return nil_id;
+        }
+        switch (std::get< role_regex >(dest)) {
+        case role_regex::LEADER: {
+            auto const leader = m_server->raft_server()->get_leader();
+            if (leader == -1) return nil_id;
+            return boost::uuids::string_generator()(id_to_str(leader));
+        } break;
+        case role_regex::ALL: {
+            return std::nullopt;
+        } break;
+        default: {
+            LOGE("Method not implemented");
+            return nil_id;
```

Comment:
Is it better to return `Result< peer_id_t >` instead of a magic sentinel that cannot hold context of the failure to resolve?

Reply:
hmm, it is better to return error with context. Its just that we will have to pass on the `std::optional< Result < peer_id_t > >` to `mesg_client`. Should be fine